// @honkit/markdown/lib/toMarkdown.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const markdown_escape_1 = __importDefault(require("markdown-escape"));
// Return N time a string
function ns(s, n) {
    return Array(n + 1).join(s);
}
/*
 * This module provides markup rules for gitbook-html
 * These rules are being used to generate SUMMARY/GLOSSARY/LANGS
 */
exports.default = {
    onText: function (text) {
        return markdown_escape_1.default(text);
    },
    onTitleStart: function (level) {
        return `${ns("#", level)} `;
    },
    onTitleEnd: function (level) {
        return this.onBL();
    },
    onParagraphStart: function () {
        return this.onSection();
    },
    onParagraphEnd: function () {
        return this.onSection();
    },
    onLinkStart: function () {
        return "[";
    },
    onLinkEnd: function (href) {
        return `](${href})`;
    },
    onListStart: function (level) {
        return "";
    },
    onListEnd: function () {
        return "";
    },
    onListItemStart: function (level) {
        return `${ns(" ", level * 4)}* `;
    },
    onListItemEnd: function () {
        return "";
    },
    onHR: function () {
        return "-----";
    },
};
