// @honkit/markup-it/lib/markup.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const parse_1 = __importDefault(require("./parse"));
const render_1 = __importDefault(require("./render"));
/**
 * Create an instance using a set of rules
 * @param {Syntax} syntax
 */
function DraftMarkup(syntax) {
    if (!(this instanceof DraftMarkup)) {
        // @ts-ignore
        return new DraftMarkup(syntax);
    }
    this.syntax = syntax;
}
/**
 * Convert a text into a parsed content
 * @param  {String} text
 * @return {ContentState}
 */
DraftMarkup.prototype.toContent = function toContent(text, options) {
    return parse_1.default(this.syntax, text, options);
};
/**
 * Convert a text into an inline parsed content
 * @param  {String} text
 * @return {List<Tokens>}
 */
DraftMarkup.prototype.toInlineContent = function toInlineContent(text) {
    // @ts-ignore
    return parse_1.default.asInline(this.syntax, text);
};
/**
 * Convert a content to text
 * @param  {ContentState} content
 * @param  {Object} options
 * @return {String}
 */
DraftMarkup.prototype.toText = function toText(content, options) {
    return render_1.default(this.syntax, content, options);
};
/**
 * Convert a content to text
 * @param  {List<Tokens>}
 * @return {String}
 */
DraftMarkup.prototype.toInlineText = function toInlineText(tokens) {
    return render_1.default.asInline(this.syntax, tokens);
};
exports.default = DraftMarkup;
