// @honkit/markup-it/lib/json/decode.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const immutable_1 = __importDefault(require("immutable"));
const token_1 = __importDefault(require("../models/token"));
const content_1 = __importDefault(require("../models/content"));
/**
 * Decode a token
 *
 * @paran {Object} json
 * @return {Token}
 */
function decodeTokenFromJSON(json) {
    // @ts-ignore
    return new token_1.default({
        type: json.type,
        text: json.text,
        raw: json.raw,
        data: json.data,
        tokens: decodeTokensFromJSON(json.tokens || []),
    });
}
/**
 * Decode a list of tokens
 *
 * @paran {Object} json
 * @return {List<Token>}
 */
function decodeTokensFromJSON(json) {
    return immutable_1.default.List(json.map(decodeTokenFromJSON));
}
/**
 * Decode a JSON into a Content
 *
 * @paran {Object} json
 * @return {Content}
 */
function decodeContentFromJSON(json) {
    return content_1.default.createFromToken(json.syntax, decodeTokenFromJSON(json.token));
}
exports.default = decodeContentFromJSON;
